File: internal/www/api/v0/libraries.go
package v0

import (
	"errors"
	"net/http"
	"os"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
	"github.com/swopcart/server/internal/services/jobs"
	"github.com/swopcart/server/internal/services/library"
)

// CreateLibraryRequest is the request body for creating a library
type CreateLibraryRequest struct {
	Name        string   `json:"name" binding:"required"`
	Description string   `json:"description"`
	PlatformID  uint     `json:"platformId" binding:"required"`
	Paths       []string `json:"paths" binding:"required,min=1"`
}

// UpdateLibraryRequest is the request body for updating a library
type UpdateLibraryRequest struct {
	Name        *string  `json:"name,omitempty"`
	Description *string  `json:"description,omitempty"`
	Paths       []string `json:"paths,omitempty"`
}

// TriggerScanRequest is the request body for manually triggering a scan
type TriggerScanRequest struct {
	LibraryID string `json:"libraryId" binding:"required"`
}

// ListLibrariesHandler returns all libraries
func (h *APIHandlers) ListLibrariesHandler(c *gin.Context) {
	// Check admin access
	sessionData := getSessionData(c)
	if !sessionData.Admin {
		respondError(c, http.StatusForbidden, ErrCodeOnlyAdmins, "Only administrators can access this resource")
		return
	}

	libraries, err := h.services.Library.ListLibraries(c.Request.Context())
	if err != nil {
		h.logger.ErrorContext(c.Request.Context(), "failed to list libraries", "error", err)
		c.Status(http.StatusInternalServerError)
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"items": libraries,
		"total": len(libraries),
	})
}

// GetLibraryHandler returns a single library
func (h *APIHandlers) GetLibraryHandler(c *gin.Context) {
	// Check admin access
	sessionData := getSessionData(c)
	if !sessionData.Admin {
		respondError(c, http.StatusForbidden, ErrCodeOnlyAdmins, "Only administrators can access this resource")
		return
	}

	libraryID := c.Param("libraryId")
	id, err := uuid.Parse(libraryID)
	if err != nil {
		respondFieldError(c, http.StatusBadRequest, ErrCodeInvalidUUID, "Invalid library ID", ".libraryId")
		return
	}

	lib, err := h.services.Library.GetLibrary(c.Request.Context(), id)
	if err != nil {
		if errors.Is(err, errors.New("library not found")) {
			respondError(c, http.StatusNotFound, ErrCodeLibraryNotFound, "Library not found")
		} else {
			h.logger.ErrorContext(c.Request.Context(), "failed to get library", "id", id, "error", err)
			c.Status(http.StatusInternalServerError)
		}
		return
	}

	c.JSON(http.StatusOK, lib)
}

// CreateLibraryHandler creates a new library
func (h *APIHandlers) CreateLibraryHandler(c *gin.Context) {
	// Check admin access
	sessionData := getSessionData(c)
	if !sessionData.Admin {
		respondError(c, http.StatusForbidden, ErrCodeOnlyAdmins, "Only administrators can access this resource")
		return
	}

	var req CreateLibraryRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		respondBindingError(c, err)
		return
	}

	// Validate paths exist
	for _, path := range req.Paths {
		info, err := os.Stat(path)
		if err != nil {
			if os.IsNotExist(err) {
				respondFieldError(c, http.StatusBadRequest, ErrCodePathNotFound, "Path does not exist: "+path, ".paths")
			} else {
				respondError(c, http.StatusBadRequest, ErrCodeInvalidLibraryPath, "Cannot access path: "+path)
			}
			return
		}
		if !info.IsDir() {
			respondFieldError(c, http.StatusBadRequest, ErrCodePathNotDirectory, "Path is not a directory: "+path, ".paths")
			return
		}
	}

	lib, err := h.services.Library.CreateLibrary(c.Request.Context(), library.CreateLibraryRequest{
		Name:        req.Name,
		Description: req.Description,
		PlatformID:  req.PlatformID,
		Paths:       req.Paths,
	})

	if err != nil {
		if errors.Is(err, errors.New("platform not found")) {
			respondFieldError(c, http.StatusBadRequest, ErrCodePlatformNotFound, "Platform not found", ".platformId")
		} else {
			h.logger.ErrorContext(c.Request.Context(), "failed to create library", "error", err)
			c.Status(http.StatusInternalServerError)
		}
		return
	}

	// Get the full library response with platform info
	fullLib, err := h.services.Library.GetLibrary(c.Request.Context(), lib.ID)
	if err != nil {
		h.logger.ErrorContext(c.Request.Context(), "failed to get created library", "error", err)
		c.Status(http.StatusInternalServerError)
		return
	}

	c.JSON(http.StatusCreated, fullLib)
}

// UpdateLibraryHandler updates a library
func (h *APIHandlers) UpdateLibraryHandler(c *gin.Context) {
	// Check admin access
	sessionData := getSessionData(c)
	if !sessionData.Admin {
		respondError(c, http.StatusForbidden, ErrCodeOnlyAdmins, "Only administrators can access this resource")
		return
	}

	libraryID := c.Param("libraryId")
	id, err := uuid.Parse(libraryID)
	if err != nil {
		respondFieldError(c, http.StatusBadRequest, ErrCodeInvalidUUID, "Invalid library ID", ".libraryId")
		return
	}

	var req UpdateLibraryRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		respondBindingError(c, err)
		return
	}

	// Validate paths if provided
	if len(req.Paths) > 0 {
		for _, path := range req.Paths {
			info, err := os.Stat(path)
			if err != nil {
				if os.IsNotExist(err) {
					respondFieldError(c, http.StatusBadRequest, ErrCodePathNotFound, "Path does not exist: "+path, ".paths")
				} else {
					respondError(c, http.StatusBadRequest, ErrCodeInvalidLibraryPath, "Cannot access path: "+path)
				}
				return
			}
			if !info.IsDir() {
				respondFieldError(c, http.StatusBadRequest, ErrCodePathNotDirectory, "Path is not a directory: "+path, ".paths")
				return
			}
		}
	}

	_, err = h.services.Library.UpdateLibrary(c.Request.Context(), id, library.UpdateLibraryRequest{
		Name:        req.Name,
		Description: req.Description,
		Paths:       req.Paths,
	})

	if err != nil {
		if errors.Is(err, errors.New("library not found")) {
			respondError(c, http.StatusNotFound, ErrCodeLibraryNotFound, "Library not found")
		} else {
			h.logger.ErrorContext(c.Request.Context(), "failed to update library", "id", id, "error", err)
			c.Status(http.StatusInternalServerError)
		}
		return
	}

	// Get the full library response with platform info
	fullLib, err := h.services.Library.GetLibrary(c.Request.Context(), id)
	if err != nil {
		h.logger.ErrorContext(c.Request.Context(), "failed to get updated library", "error", err)
		c.Status(http.StatusInternalServerError)
		return
	}

	c.JSON(http.StatusOK, fullLib)
}

// DeleteLibraryHandler deletes a library
func (h *APIHandlers) DeleteLibraryHandler(c *gin.Context) {
	// Check admin access
	sessionData := getSessionData(c)
	if !sessionData.Admin {
		respondError(c, http.StatusForbidden, ErrCodeOnlyAdmins, "Only administrators can access this resource")
		return
	}

	libraryID := c.Param("libraryId")
	id, err := uuid.Parse(libraryID)
	if err != nil {
		respondFieldError(c, http.StatusBadRequest, ErrCodeInvalidUUID, "Invalid library ID", ".libraryId")
		return
	}

	err = h.services.Library.DeleteLibrary(c.Request.Context(), id)
	if err != nil {
		h.logger.ErrorContext(c.Request.Context(), "failed to delete library", "id", id, "error", err)
		c.Status(http.StatusInternalServerError)
		return
	}

	c.JSON(http.StatusOK, gin.H{"success": true})
}

// TriggerLibraryScanHandler manually triggers a library scan
func (h *APIHandlers) TriggerLibraryScanHandler(c *gin.Context) {
	// Check admin access
	sessionData := getSessionData(c)
	if !sessionData.Admin {
		respondError(c, http.StatusForbidden, ErrCodeOnlyAdmins, "Only administrators can access this resource")
		return
	}

	libraryID := c.Param("libraryId")
	id, err := uuid.Parse(libraryID)
	if err != nil {
		respondFieldError(c, http.StatusBadRequest, ErrCodeInvalidUUID, "Invalid library ID", ".libraryId")
		return
	}

	// Check if library exists
	lib, err := h.services.Library.GetLibrary(c.Request.Context(), id)
	if err != nil {
		respondError(c, http.StatusNotFound, ErrCodeLibraryNotFound, "Library not found")
		return
	}

	// Check if scan already in progress
	if lib.CurrentScanJobID != nil {
		respondError(c, http.StatusConflict, ErrCodeScanAlreadyInProgress, "Scan already in progress for this library")
		return
	}

	// Enqueue scan job with parameters
	executionID, err := h.services.Jobs.EnqueueJob(
		c.Request.Context(),
		"library.scan",
		jobs.WithParameters(jobs.Params{"libraries": id.String()}),
	)

	if err != nil {
		h.logger.ErrorContext(c.Request.Context(), "failed to enqueue scan job", "id", id, "error", err)
		c.Status(http.StatusInternalServerError)
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"executionId": executionID.String(),
		"status":      "pending",
	})
}

File: frontend/src/lib/api/libraries.ts
/**
 * Libraries API
 *
 * Handles library management (admin-only), including CRUD operations and scan triggering.
 * Endpoints under /api/v0/libraries/*
 */

import { z } from "zod";
import { ApiError, fetchWithAuth } from "./client";
import { LibrarySchema, type Library } from "./types";

// Re-export types for external use
export type { Library };

// Create library request
export const CreateLibraryRequestSchema = z.object({
  name: z.string().min(1),
  description: z.string().optional(),
  platformId: z.number(),
  paths: z.array(z.string()).min(1),
});

export type CreateLibraryRequest = z.infer<typeof CreateLibraryRequestSchema>;

// Update library request
export const UpdateLibraryRequestSchema = z.object({
  name: z.string().optional(),
  description: z.string().optional(),
  paths: z.array(z.string()).optional(),
});

export type UpdateLibraryRequest = z.infer<typeof UpdateLibraryRequestSchema>;

// Scan response
export const TriggerScanResponseSchema = z.object({
  executionUuid: z.string(),
  status: z.string(),
});

export type TriggerScanResponse = z.infer<typeof TriggerScanResponseSchema>;

// API functions

/**
 * List all libraries (admin-only)
 */
export async function listLibraries(): Promise<Library[]> {
  const response = await fetchWithAuth("/libraries/", {
    method: "GET",
  });

  if (!response.ok) {
    const body = await response.json();
    throw new ApiError(response.status, body);
  }

  const data = await response.json();
  return z.array(LibrarySchema).parse(data.items);
}

/**
 * Get a single library by ID (admin-only)
 */
export async function getLibrary(libraryId: string): Promise<Library> {
  const response = await fetchWithAuth(`/libraries/${libraryId}`, {
    method: "GET",
  });

  if (!response.ok) {
    const body = await response.json();
    throw new ApiError(response.status, body);
  }

  return LibrarySchema.parse(await response.json());
}

/**
 * Create a new library (admin-only)
 */
export async function createLibrary(
  request: CreateLibraryRequest,
): Promise<Library> {
  const response = await fetchWithAuth("/libraries/", {
    method: "POST",
    body: JSON.stringify({
      name: request.name,
      description: request.description || "",
      platformId: request.platformId,
      paths: request.paths,
    }),
  });

  if (!response.ok) {
    const body = await response.json();
    throw new ApiError(response.status, body);
  }

  return LibrarySchema.parse(await response.json());
}

/**
 * Update a library (admin-only)
 */
export async function updateLibrary(
  libraryId: string,
  request: UpdateLibraryRequest,
): Promise<Library> {
  const response = await fetchWithAuth(`/libraries/${libraryId}`, {
    method: "PATCH",
    body: JSON.stringify(request),
  });

  if (!response.ok) {
    const body = await response.json();
    throw new ApiError(response.status, body);
  }

  return LibrarySchema.parse(await response.json());
}

/**
 * Delete a library (admin-only)
 */
export async function deleteLibrary(libraryId: string): Promise<void> {
  const response = await fetchWithAuth(`/libraries/${libraryId}`, {
    method: "DELETE",
  });

  if (!response.ok) {
    const body = await response.json();
    throw new ApiError(response.status, body);
  }
}

/**
 * Trigger a library scan (admin-only)
 */
export async function triggerLibraryScan(
  libraryId: string,
): Promise<TriggerScanResponse> {
  const response = await fetchWithAuth(`/libraries/${libraryId}/scan`, {
    method: "POST",
    body: JSON.stringify({}),
  });

  if (!response.ok) {
    const body = await response.json();
    throw new ApiError(response.status, body);
  }

  return TriggerScanResponseSchema.parse(await response.json());
}
